src/ILCompiler.Compiler/src/Compiler/InteropStubManager.cs

```diff
@@ -101,32 +105,172 @@ internal TypeDesc GetInlineArrayType(InlineArrayCandidate candidate)
+            //    {
+            //        AddDependenciesDuePInvokeStruct(ref dependencies, factory, type);
+            //    }
+            //}
+        }
+
+
+        /// <summary>
+        /// For Marshal generic APIs(eg. Marshal.StructureToPtr<T>, GetFunctionPointerForDelegate) we add
+        /// the generic parameter as dependencies so that we can generate runtime data for them
+        /// </summary>
+        public void AddMarshalAPIsGenericDependencies(ref DependencyList dependencies, NodeFactory factory, MethodDesc method)
+        {
+            Debug.Assert(method.HasInstantiation);
+
+            TypeDesc owningType = method.OwningType;
+            MetadataType metadataType = owningType as MetadataType;
+            if (metadataType != null && metadataType.Module == _interopModule)
+            {
+                if (metadataType.Name == "Marshal" && metadataType.Namespace == "System.Runtime.InteropServices")
+                {
+                    string methodName = method.Name;
+                    if (methodName == "GetFunctionPointerForDelegate" ||
+                        methodName == "GetDelegateForFunctionPointer" ||
+                        methodName == "PtrToStructure" ||
+                        methodName == "StructureToPtr" ||
+                        methodName == "SizeOf" ||
+                        methodName == "OffsetOf")
+                    {
+                        foreach (TypeDesc type in method.Instantiation)
+                        {
+                            AddPInvokeParameterDependencies(ref dependencies, factory, type);
+                        }
+                    }
+                }
+            }
+        }
+
+        public void AddPInvokeParameterDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc parameter)
+        {
+            AddDependenciesDueToPInvokeDelegate(ref dependencies, factory, parameter);
+            AddDependenciesDueToPInvokeStruct(ref dependencies, factory, parameter);
+        }
+
+        public void AddDependenciesDueToPInvokeDelegate(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
+        {
+            if (type.IsDelegate)
+            {
+                dependencies.Add(factory.NecessaryTypeSymbol(type), "Delegate Marshalling Stub");
+
+                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetOpenStaticDelegateMarshallingStub(type)), "Delegate Marshalling Stub");
+                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetClosedDelegateMarshallingStub(type)), "Delegate Marshalling Stub");
+                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetForwardDelegateCreationStub(type)), "Delegate Marshalling Stub");
+            }
+        }
+
+        public void AddDependenciesDueToPInvokeStruct(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
```

Comment:
@jkotas As previously discussed, this is a short term solution to get the basics working for Marshal API before we start looking at more longer term IL linker based solution.